subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/deps/ClassSetAnalysis.java

```diff
@@ -115,180 +110,101 @@ public ClassSetDiff findChangesSince(ClassSetAnalysis other) {
-
-        Set<String> privateResultClasses = new HashSet<>();
-        Set<String> accessibleResultClasses = new HashSet<>();
-        Set<GeneratedResource> resultResources = new HashSet<>(resourcesDependingOnAllOthers);
-        processDependentClasses(new HashSet<>(), privateResultClasses, accessibleResultClasses, resultResources, deps.getPrivateDependentClasses(), dependents);
-        processDependentClasses(new HashSet<>(), privateResultClasses, accessibleResultClasses, resultResources, Collections.emptySet(), classesDependingOnAllOthers);
-        processDependentClasses(new HashSet<>(), privateResultClasses, accessibleResultClasses, resultResources, constantDeps.getPrivateDependentClasses(), constantDeps.getAccessibleDependentClasses());
-        accessibleResultClasses.remove(className);
-        privateResultClasses.remove(className);
-
-        return DependentsSet.dependents(privateResultClasses, accessibleResultClasses, resultResources);
-    }
-
-    public Set<String> getTypesToReprocess() {
-        // Because of https://github.com/gradle/gradle/issues/13767 and
-        // https://github.com/gradle/gradle/issues/15009 it is possible
-        // that the types to reprocess are actually generated types
-        // so when we see a type to reprocess we need to track what
-        // actually generated this type, not use it directly!
-        return annotationProcessingData.getAggregatedTypes()
-            .stream()
-            .map(annotationProcessingData::getOriginOf)
-            .collect(Collectors.toSet());
-    }
+        Set<String> privateDependents = new HashSet<>();
+        Set<String> accessibleDependents = new HashSet<>();
+        Set<GeneratedResource> dependentResources = new HashSet<>();
+        Set<String> visited = new HashSet<>();
+        Deque<String> remaining = new ArrayDeque<>();
+        remaining.add(className);
 
-    public boolean isDependencyToAll(String className) {
-        return classAnalysis.getDependents(className).isDependencyToAll();
-    }
-
-    /**
-     * Accumulate dependent classes and resources. Dependent classes discovered can themselves be used to query
-     * further dependents, while resources are just data accumulated along the way. Recurses for classes that
-     * are "publicly accessbile", i.e. classes that are not just used privately in a class.
-     */
-    private void processDependentClasses(Set<String> visitedClasses,
-                                         Set<String> privateResultClasses,
-                                         Set<String> accessibleResultClasses,
-                                         Set<GeneratedResource> resultResources,
-                                         Iterable<String> privateDependentClasses,
-                                         Iterable<String> accessibleDependentClasses) {
-
-        for (String privateDependentClass : privateDependentClasses) {
-            if (!visitedClasses.add(privateDependentClass)) {
```

Comment:
This broke any dependents that were both private and accessible - They were processed here first and then ignored in the transitive dependency analysis further down, because they were already in `visitedClasses`.